pytorch_lightning/utilities/trainer_cli.py

```diff
@@ -0,0 +1,221 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+from typing import Type
+from jsonargparse import ArgumentParser, ActionConfigFile
+from pytorch_lightning import Trainer, LightningModule, LightningDataModule
+from pytorch_lightning.callbacks import Callback
+
+
+class LightningArgumentParser(ArgumentParser):
+    """Extension of jsonargparse's ArgumentParser for pythorch-lightning"""
+
+    def __init__(
+        self,
+        *args,
+        parse_as_dict: bool = True,
+        **kwargs
+    ):
+        """Initialize argument parser that supports configuration file input"""
+        super().__init__(*args, parse_as_dict=parse_as_dict, **kwargs)
+        self.add_argument(
+            '--config',
+            action=ActionConfigFile,
+            help='Path to a configuration file in json or yaml format.'
+        )
+
+    def add_trainer_args(
+        self,
+        trainer_class: Type[Trainer] = Trainer,
+        nested_key: str = 'trainer'
+    ):
+        """
+        Adds arguments from a trainer class to a nested key of the parser
+
+        Args:
+            trainer_class: Optional extension of the Trainer class.
+            nested_key: Name of the nested namespace to store arguments.
+        """
+        assert issubclass(trainer_class, Trainer)
+        self.add_class_arguments(trainer_class, nested_key)
+
+    def add_module_args(
+        self,
+        module_class: Type[LightningModule],
+        nested_key: str = 'module'
+    ):
+        """
+        Adds arguments from a module class to a nested key of the parser
+
+        Args:
+            module_class: A LightningModule class.
+            nested_key: Name of the nested namespace to store arguments.
+        """
+        assert issubclass(module_class, LightningModule)
+        self.add_class_arguments(module_class, nested_key)
+
+    def add_datamodule_args(
+        self,
+        datamodule_class: Type[LightningDataModule],
+        nested_key: str = 'data'
+    ):
+        """
+        Adds arguments from a datamodule class to a nested key of the parser
+
+        Args:
+            datamodule_class: A LightningDataModule class.
+            nested_key: Name of the nested namespace to store arguments.
+        """
+        assert issubclass(datamodule_class, LightningDataModule)
+        self.add_class_arguments(datamodule_class, nested_key)
+
+
+class SaveConfigCallback(Callback):
+    """Saves a TrainerCli config to the log_dir when training starts"""
+
+    def __init__(self, parser, config):
+        self.config_dump = parser.dump(config, skip_none=False)
+
+    def on_train_start(self, trainer, pl_module):
+        config_path = os.path.join(trainer.logger.log_dir, 'config.yaml')
+        with open(config_path, 'w') as outstream:
+            outstream.write(self.config_dump)
+
+
+class TrainerCli:
```

Comment:
```suggestion
class LightningCLI:
```
the trainer in the name is a bit confusing as it takes Model and datamodule

Reply:
I like it. Just changed it.